src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java

```diff
@@ -285,11 +309,22 @@ public long count(Query query, Class<?> clazz) {
 		int c = 0;
 		Iterator<Class<?>> it1 = classes.iterator();
 		for (Query query : queries) {
-			res.add(elasticsearchConverter.read(it1.next(), SearchDocumentResponse.from(items[c++].getResponse())));
+			Class entityClass = it1.next();
+
+			SearchDocumentResponseCallback<SearchHits<?>> callback = readSearchDocumentResponseCallback(
+					entityClass, index);
+
+			SearchResponse response = items[c++].getResponse();
+			res.add(callback.doWith(SearchDocumentResponse.from(response)));
 		}
 		return res;
 	}
 
+	private <T> SearchDocumentResponseCallback<SearchHits<T>> readSearchDocumentResponseCallback(Class<T> entityClass,
```

Comment:
can we cache these for (entityClass, index)? to prevent new creation?
 Are the callbacks then threadsafe, if we reuse cached instances?

Reply:
What's the point? `ReadSearchDocumentResponseCallback` is created once per query. A query execution takes orders of magnitide more time than callback object allocation (because it involves network I/O and, probably, disk I/O on the server).
In terms of garbage, one or two objects per query (given that each query may produce dozens, hundreds or more entities) does not seem to be a high price as well.

Reply:
forget about that caching idea.

`readSearchDocumentResponseCallback(...)` is a private method that's a one-liner and is only called at one place in the second _multiSearch_ method. I think we could inline this method.

Reply:
The only reason I felt this method was needed was to capture the type variable. Without this, it would require to use even more raw types. Are you ok with this?

Reply:
If I inline this I get:

```
for (Query query : queries) {
	Class entityClass = it1.next();

	SearchDocumentResponseCallback<SearchHits<?>> callback = new ReadSearchDocumentResponseCallback<>(entityClass, index);

	SearchResponse response = items[c++].getResponse();
	res.add(callback.doWith(SearchDocumentResponse.from(response)));
}
```
same as in the other `multiSearch` method above this one.

Reply:
Oh, sure, I introduced this method when 2 separate callbacks were created in it with the same type parameter. Now it is not needed, I've inlined it. Thanks.